toad.8
.\"
.\" Copyright (c) 2016 Antoine Jacoutot <ajacoutot@openbsd.org>
.\" Copyright (c) 2013 M:tier Ltd.
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\"
.Dd $Mdocdate: December 18 2016 $
.Dt TOAD 8
.Os
.Sh NAME
.Nm toad
.Nd
device automounter for
.Xr hotplugd 8
.Sh SYNOPSIS
.Nm
.Bk -words
.Ar action
.Ar devclass
.Ar devname
.Ek
.Sh DESCRIPTION
.Nm
(Toad Opens All Devices)
is a utility meant to be started from the
.Xr hotplugd 8
.Em attach
and
.Em detach
scripts.
It will try to mount all partitions found on the device under
.Pa /run/media/${USER}/device .
Where ${USER} is the active user
.Em login
name and
.Em device
is the type of the device,
.Em usb
or
.Em cd ,
followed by its number (from
.Em 0
to
.Em 9 ) .
This follows the
.Em udev
hierarchy in Linux which allows interaction with GLib/GIO's GUnixMount.
.Pp
Users outside the default UID range (1000...60000) are ignored (nothing will
be done).
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Ar action
Can be
.Ar attach
(for device attachment)
or
.Ar detach
(for device detachment).
.It Ar devclass
Integer corresponding to the device class as described in the
.Aq Pa sys/device.h
header file,  e.g.\&
.Pa 2
for disk drives.
.It Ar devname
Device name as listed in
.Xr MAKEDEV 8 ,
and the unit number, e.g.\&
.Pa sd1 .
.El
.Pp
Detection of the currently active user is done using ConsoleKit and DBus,
.Nm
will not do anything unless these are properly setup and running.
Obviously,
.Xr hotplugd 8
must be running as well.
.Xr rcctl 8
can be used to enable and start
.Xr hotplugd 8
and
.Xr dbus-daemon 1 :
.Bd -literal -offset indent
.Ev rcctl enable hotplugd messagebus
.Ev rcctl start hotplugd messagebus
.Ed
.Sh EXAMPLES
This sample code can be used for both the
.Em attach
and
.Em detach
scripts:
.Bd -literal -offset indent
#!/bin/sh

DEVCLASS=$1
DEVNAME=$2

case $DEVCLASS in
2)
	/usr/local/libexec/toad ${0##*/} $DEVCLASS $DEVNAME
	;;
esac
.Ed
.Sh SEE ALSO
.Xr dbus-daemon 1 ,
.Xr hotplugd 8
.Sh AUTHORS
.An Antoine Jacoutot Aq Mt antoine@mtier.org


ChangeLog
v1.9 - NOT-RELEASED-YET
-----------------------------------

v1.8 - Sun Feb 10 15:03:09 CET 2019
-----------------------------------
- only works with regular users (uid 1000...60000)
- add support for ejecting cd-rom
- remove support for firmware installation, it can easily be done using
  fw_update(1) in the hotplugd(8) attach script

v1.7 - Wed Jan 23 04:37:24 CET 2019
-----------------------------------
- big rewrite to use pkexec(1), needs glib2 >=2.58.3p0 (OpenBSD package)
- add net device to the attach script sample

v1.6 - Thu Feb 25 13:53:42 CET 2016
-----------------------------------
- only react on explicit hotplug(4) events

v1.5 - Thu Feb 25 13:28:36 CET 2016
-----------------------------------
- add support for fw_update(1); from Robert Nagy (robert@mtier.org)

v1.4 - Wed Dec 23 15:09:54 CET 2015
-----------------------------------
- drop support for opening the default file manager
- fix output when running from the cli

v1.3 - Mon Jul 20 02:49:04 CET 2015
-----------------------------------
- display an error message if we cannot mount a device
- properly detect our xmessage compatible command

v1.2 - Thu Apr 23 15:49:36 CET 2015
-----------------------------------
- rc.d(8) script for dbus-daemon(1) was renamed (Adam Wolk)
- tweak error messages
- fix typo

v1.1 - Sun Mar  2 17:24:40 CET 2014
-----------------------------------
- make toad(8) more resilient to weird corner cases

v1.0 - Sun Mar  2 13:48:27 CET 2014
-----------------------------------
- drop all secondary groups inherited from root when we fork and setgid
- mount point and device ownership are now created and set by the child
- add a call to restore_dev in broom_sweep
- do not try to mount 'c': not our job to deal with broken disklabels
- make toad(8) more resilient, sanitize all return codes
- extend documentation a bit after users'feedback
- do nothing if uid is lower than 1000 (OpenBSD default UID_MIN)
- display is not optional since we have no ConsoleKit PAM support on OpenBSD

v0.9 - Fri Dec 27 10:05:28 CET 2013
-----------------------------------
- Add UDF support
- Add a sample attach+detach script
- Properly check for devices locked in read-only mode
- Fix zenity window height and width

v0.8 - Wed Dec 11 12:54:22 CET 2013
-----------------------------------
- Ignore softraid(4) crypto disk attachement
- Only change the device owner, not the group
- Fix ejecting optical media, gio passes the block device path as
  argument to eject(1), not the raw one

v0.7 - Mon Dec  9 12:32:09 CET 2013
-----------------------------------
- toadd: fix path to toad(8)

v0.6 - Mon Dec  9 10:08:37 CET 2013
-----------------------------------
- Exit and clean when no partition is found on the device
- Only attempt to mount the 'c' partition if no other partition is
  found and the label is MSDOS
- Try to mount R/O if the initial mount command failed (and warn the user)

v0.5 - Thu Dec  5 15:19:24 CET 2013
-----------------------------------
- Fix compilation with WARNINGS set to "Yes"
  Giovanni Bechis (giovanni@openbsd.org)

v0.4 - Wed Dec  4 14:35:27 CET 2013
-----------------------------------
- do not hardcode path to toad(8)
- stricter permissions on mount path

v0.3 - Tue Oct  8 18:43:31 CET 2013
-----------------------------------
- fix install target
- minor cleanups

v0.2 - Thu Oct  4 14:18:58 UTC 2013
-----------------------------------
- toadd: new option `-w' to specify the optical disk polling interval
- toadd: new option `-d' to run in foreground with debug output

v0.1 - Thu Oct  3 14:18:58 UTC 2013
-----------------------------------
- Initial import of toad-0.1


toad.pl
#!/usr/bin/perl
#
# Copyright (c) 2016, 2019 Antoine Jacoutot <ajacoutot@openbsd.org>
# Copyright (c) 2013, 2014, 2015, 2016 M:tier Ltd.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
# Author: Antoine Jacoutot <antoine@mtier.org>

use strict;
use warnings;

use File::Path qw(make_path);
use Net::DBus;
use User::pwent qw(:FIELDS);

if ( $< != 0 ) {
	print "need root privileges\n";
	exit (1);
}

sub usage {
	print "usage: $0 attach|detach devclass device\n";
	exit (1);
}

if (@ARGV < 3) { usage (); }

my ($action, $devclass, $devname) = @ARGV;
my ($login, $uid, $gid, $display) = get_active_user_info ();
my $mounttop = '/run';
my $mountbase = "$mounttop/media";
my $mountopts = 'nodev,nosuid,noexec';
my $devtype;
my $devmax;
my $pkrulebase = "/etc/polkit-1/rules.d/45-toad-$login";

sub broom_sweep {
	my $is_busy;
	my @tryrm;
	my @cfd;

	unlink glob "$pkrulebase-$devname?.rules";

	if (-d $mountbase && $mountbase ne '/') {
		opendir (TOP, $mountbase) or return;
		while (my $file = readdir (TOP)) {
			next if ($file =~ m/^\./);
			next unless (-d "$mountbase/$file");
			push @cfd, "$mountbase/$file";
			opendir (SUB, "$mountbase/$file") or die "cannot open $mountbase/$file: $!";
			while (my $subfile = readdir (SUB)) {
				next if ($subfile =~ m/^\./);
				next unless (-d "$mountbase/$file/$subfile");
				if ((stat ("$mountbase/$file/$subfile"))[1] != 2) {
					push @cfd, "$mountbase/$file/$subfile";
				} else {
					# subdir is mounted, so don't try to
					# remove parent
					$is_busy = 1;
					my $i = 0;
					my $c = scalar @cfd;
					while ($cfd[$i]) {
						$i++ until $cfd[$i] eq "$mountbase/$file" or $i==$c;
						splice (@cfd, $i, 1);
					}
				}
			}
			closedir (SUB);
		}
		closedir (TOP);

		@tryrm = reverse sort @cfd;

		foreach my $rmfile (@tryrm) {
			rmdir ($rmfile);
		}

		# nothing is mounted, so remove the hierarchy
		unless ($is_busy) {
			rmdir ($mountbase);
			rmdir ($mounttop);
		}
	}
}

sub create_hier {
	make_path ($mountbase, {owner=>0, group=>0, mode=>0755});
	chown 0, 0, $mounttop, $mountbase;
	chmod 0755, $mounttop, $mountbase;
}

sub create_mount_point {
	my $devnum = shift;

	create_hier ();

	make_path ("$mountbase/$login/$devtype$devnum", {owner=>$uid, group=>$gid, mode=>0700});
	chown $uid, $gid, "$mountbase/$login", "$mountbase/$login/$devtype$devnum";
	chmod 0700, "$mountbase/$login", "$mountbase/$login/$devtype$devnum";
}

sub create_pkrule {
	my($devname, $devnum, $part) = @_;
	my $pkrule = "$pkrulebase-$devname$part.rules";

	unless(open PKRULE, '>'.$pkrule) {
		die "Unable to create $pkrule\n";
	}

	print PKRULE "polkit.addRule(function(action, subject) {\n";
	print PKRULE "  if (action.id == \"org.freedesktop.policykit.exec\" &&\n";
	print PKRULE "    action.lookup(\"program\") == \"/sbin/umount\" &&\n";
	print PKRULE "    action.lookup(\"command_line\") == \"/sbin/umount $mountbase/$login/$devtype$devnum\") {\n";
	print PKRULE "    if (subject.local && subject.active && subject.user == \"$login\") {\n";
	print PKRULE "      return polkit.Result.YES;\n";
	print PKRULE "    }\n";
	print PKRULE "  }\n";
	print PKRULE "});\n";

	if ($devtype eq 'cd') {
		print PKRULE "polkit.addRule(function(action, subject) {\n";
		print PKRULE "  if (action.id == \"org.freedesktop.policykit.exec\" &&\n";
		print PKRULE "    action.lookup(\"program\") == \"/bin/eject\" &&\n";
		print PKRULE "    action.lookup(\"command_line\") == \"/bin/eject /dev/$devname$part\") {\n";
		print PKRULE "    if (subject.local && subject.active && subject.user == \"$login\") {\n";
		print PKRULE "      return polkit.Result.YES;\n";
		print PKRULE "    }\n";
		print PKRULE "  }\n";
		print PKRULE "});\n";
	}

	close PKRULE;
}

sub gdbus_call {
	my($action, $args) = @_;
	my $gdbus_call = "/usr/local/bin/gdbus call";

	my $pid = fork ();
	if (!defined ($pid)) {
		die "could not fork: $!";
	} elsif ($pid) {
		if (waitpid ($pid, 0) > 0) {
			if ($? >> 8 ne 0) {
				return (1);
			}
		}
	} else {
		$( = $) = "$gid $gid";
		$< = $> = $uid;

		$ENV{"DISPLAY"} = $display;
		if ($action eq 'notify') {
			system ("$gdbus_call -e -d org.freedesktop.Notifications -o /org/freedesktop/Notifications -m org.freedesktop.Notifications.Notify toad 42 drive-harddisk-usb \"Toad\" \"$args\" [] {} 5000 >/dev/null");
		} elsif ($action eq 'open-fm') {
			system ("$gdbus_call -e -d org.freedesktop.FileManager1 -o /org/freedesktop/FileManager1 -m org.freedesktop.FileManager1.ShowFolders '[\"file://$args\"]' \"\" >/dev/null");
		}
		# exit the child
		exit (0);
	}
}

sub get_active_user_info {
	my $system_bus = Net::DBus->system;
	my $ck_service = $system_bus->get_service ('org.freedesktop.ConsoleKit');
	my $ck_manager = $ck_service->get_object ('/org/freedesktop/ConsoleKit/Manager');

	for my $session_id (@{$ck_manager->GetSessions ()}) {
		my $ck_session = $ck_service->get_object ($session_id);
		next unless $ck_session->IsActive ();

		my $uid = $ck_session->GetUnixUser ();
		getpwuid ($uid) || die "no $uid user: $!";
		next unless ($uid >= 1000 && $uid <= 60000);

		my $display = $ck_session->GetX11Display ();
		next unless length ($display);

		my $gid = $pw_gid;
		my $login = $pw_name;

		return ($login, $uid, $gid, $display);
	}
}

sub get_mount_point {
	my $devnum;

	for ($devnum = 0; $devnum < $devmax; $devnum = $devnum + 1) {
		my $mounts = `/sbin/mount | /usr/bin/grep $mountbase/$login/$devtype$devnum`;
		next unless ($mounts eq '');
		last;
	}
	return ($devnum);
}

sub get_parts {
	my @parts;
	my @supportedfs = ('MSDOS', 'NTFS', '4.2BSD', 'ext2fs', 'ISO9660', 'UDF');

	foreach my $fs (@supportedfs) {
		my $fsmatch = `/sbin/disklabel $devname 2>/dev/null | /usr/bin/grep " $fs "`;
		while ($fsmatch =~ /([^\n]+)\n?/g) {
			my @part = split /:/, $1;
			push (@parts, $part[0]);
		}
	}

	return (@parts);
}

sub mount_device {
	my @allparts;
	my @parts;

	# ignore softraid(4) crypto device attachment
	my $sr_crypto = `/sbin/disklabel $devname 2>/dev/null | grep "SR CRYPTO"`;
	if ($sr_crypto) {
		return (0);
	}

	if ($devtype eq 'cd') {
		@parts = 'a';
	} else {
		@allparts = get_parts ();
		foreach my $part (@allparts) {
			if ($part !~ 'c$') {
				push @parts, $part;
			}
		}
	}

	unless (@parts) {
		gdbus_call ("notify", "No supported partition found on device $devname!");
		return (0);
	}

	foreach my $part (@parts) {
		$part =~ s/^\s+//;
		my $device = "/dev/$devname$part";
		my $devnum = get_mount_point ();

		create_mount_point ($devnum);
		create_pkrule ($devname, $devnum, $part);

		my $mountrw = `/sbin/mount -o $mountopts $device $mountbase/$login/$devtype$devnum 2>&1`;
		if (length ($mountrw) != 0) {
			system ("/sbin/mount -o $mountopts,ro $device $mountbase/$login/$devtype$devnum");
			unless ($? == 0) {
				gdbus_call ("notify", "Cannot mount $device!");
				broom_sweep ();
				next;
			}
			unless ($mountrw =~ /Permission denied/) {
				gdbus_call ("notify", "Unclean filesystem on device $device, mounting read-only!");
			}
		}

		gdbus_call ("open-fm", "$mountbase/$login/$devtype$devnum");
	}
}

if ($devclass == 2) {
	$devtype = 'usb';
	$devmax = 10;
} elsif ($devclass == 9) {
	$devtype = 'cd';
	$devmax = 2;
} else {
	print "device type not supported\n";
	exit (1);
}

if ($action eq 'attach') {
	if (!defined($login) || !defined($uid) || !defined($gid)) {
		print "ConsoleKit: user does not own the active session\n";
		exit (1);
	}
	if ($devtype eq 'cd' || $devtype eq 'usb') { mount_device (); }
} elsif ($action eq 'detach') {
	if ($devtype eq 'cd' || $devtype eq 'usb') { broom_sweep (); }
} else { usage (); }
